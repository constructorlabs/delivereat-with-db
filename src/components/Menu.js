import React from "react";
import Basket from "./Basket";
import MoreInfo from "./MoreInfo";

import "../styles/Menu.scss";

class Menu extends React.Component {
  constructor() {
    super();
    this.state = { basket: [] };
    this.getCourse = this.getCourse.bind(this);
    this.basketReceiveRemove = this.basketReceiveRemove.bind(this);
    this.basketReceiveAdd = this.basketReceiveAdd.bind(this);
    this.calculate = this.calculate.bind(this);
  }

  

  basketReceiveAdd(id) {
    this.props.basketStatus();

    const existingItem = this.state.basket.find(element => {
      return element.menuItemId === id;
    });
    if (existingItem) {
      this.setState(
        {
          basket: this.state.basket.map(dish => {
            if (dish == existingItem) {
              return Object.assign({}, dish, { quantity: dish.quantity + 1 });
            } else {
              return dish;
            }
          })
        },
        () => console.log(this.state.basket)
      );
    } else {
      const item = { menuItemId: id, quantity: 1 };
      this.setState(
        {
          basket: this.state.basket.concat([item])
        },
        () => console.log(this.state.basket)
      );
    }
    this.calculate()
  }

  basketReceiveRemove(id) {
    const existingItem = this.state.basket.find(element => {
      return element.menuItemId === id;
    });
    if (existingItem) {
      this.setState(
        {
          basket: this.state.basket.map(dish => {
            if (dish == existingItem) {
              return Object.assign({}, dish, { quantity: dish.quantity - 1 });
            } else {
              return dish;
            }
          })
        },
        () => console.log(this.state.basket)
      );
    } else {
      const item = { menuItemId: id, quantity: 0 };
      this.setState(
        {
          basket: this.state.basket.concat([item])
        },
        () => console.log(this.state.basket)
      );
    }
    this.calculate();
    
  }


  getCourse(course) {
    return this.props.menu
      .filter(item => item.type === course)
      .map(foodItem => {
        // console.log(this.state.basket);
        // let counter = typeof this.state.basket[foodItem.id].quantity === undefined ? 0 : this.state.basket[foodItem.id].quantity;
        //  const basketQuant =
        // let counter = typeof this.state.basket[foodItem.id] === undefined ? 0: this.state.basket[0].quantity;
        let counter = typeof this.state.basket[foodItem.id] === 'undefined' ? 0: this.state.basket[foodItem.id].quantity;
        // console.log(this.state.basket[0].quantity);

        
        console.log(counter)
        // foodItem.id
        // const inBasket = order.items.length && order.items.map(item => item[0]).includes(menuItem.id);
        // const count = inBasket ? order.items.filter(item => item[0] === menuItem.id)[0][1] : 0;

        return (
          <li key={foodItem.id}>
            <div className="menuitems">
              <div className="menuitems__row">
                <h2>{foodItem.name}</h2>
                <div className="menuitems__order">

                  <button
                    className="orderbutton__plus"
                    onClick={() => this.basketReceiveAdd(foodItem.id)}
                  >
                    +
                  </button>
                  {/* <h3>0</h3> */}
                  <h3>{counter}</h3>

                  <button
                    className="orderbutton__minus"
                    onClick={() => this.basketReceiveRemove(foodItem.id)}
                  >
                    -
                  </button>
                </div>
              </div>
              <h3 className="menuitems__info">{foodItem.headline}</h3>
              <h2>{`Â£${foodItem.price}`}</h2>
            </div>
            <hr />
          </li>
        );
      });
  }

  calculate() {
    let delivery = 2.5;
    let calc = this.state.basket.map(order => {
      console.log("hi there", { order });
      console.log(this.props.menu[order.menuItemId]);
      // return parseInt(this.props.menu[order.menuItemId].price, 10 * order.quantity);
      return this.props.menu[order.menuItemId].price * order.quantity;
    });
    let calcTotal = calc.reduce((total, amount) => total + amount);

    //

    console.log("calculate", calc, calcTotal);
  }
  
  // calculate() {
  //   const menu = this.props.menu;
  //   console.log(menu);
  //   const calcTotal = this.state.basket
  //     .map(
  //       order =>
  //         menu.find(menuItem => menuItem.id === order[0]).price * order[1]
  //     )
  //     .reduce((total, amount) => total + amount);
  //   return calcTotal;
  //   console.log(calcTotal);
  // }

  render() {
    return (
      <div>
        <ul className="menu">
          <h1>BRUNCH</h1>
          {this.getCourse("brunch")}

          <h1>DRINKS</h1>
          {this.getCourse("drinks")}

          <h1>DESSERT</h1>
          {this.getCourse("dessert")}
        </ul>
        <MoreInfo
          basket={this.state.basket}
          menu={this.props.menu}
          />
        
        <Basket
          basket={this.state.basket}
          menu={this.props.menu}
          receiveOrder={this.props.receiveOrder}
          basketReceiveRemove={this.basketReceiveRemove}
          basketReceiveAdd={this.basketReceiveAdd}
          calculate={this.calculate}
          closeBasket={this.props.closeBasket}
          showBasket={this.props.showBasket}
          basketStatus={this.basketStatus}
        />
      </div>
    );
  }
}

export default Menu;
