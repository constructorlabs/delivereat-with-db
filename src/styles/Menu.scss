@import "defaults";

.menu {
  padding: 0;
}

.menuitems {
  display: flex;
  flex-direction: column;
  background-color: white;

  &__row {
    display: flex;
    justify-content: space-between;
  }

  &__order {
    display: flex;
    flex-direction: row;
    justify-content: space-around;
    align-items: baseline;
    // display: inline;
  }

  &__info {
    width: 70%;
  }
}

hr {
  margin: 0;
  padding: 0;
}

// .menuitems__details {
//   display: flex;
//   //   flex-direction: column;
//   align-items: flex-start;
// }
