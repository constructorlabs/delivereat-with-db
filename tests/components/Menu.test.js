import React from 'react';
import renderer from 'react-test-renderer';
import Menu from '../../src/components/Menu';

const stage = 'basket';
const order = { id: 10, items: [[1, 1], [2, 3]] };
const menu = [
  {
    id: 1,
    name: 'Espresso',
    description: 'Single shot, 30ml',
    price: '2.00',
  },
  {
    id: 2,
    name: 'Ristretto',
    description: 'Concentrated espresso, 22ml',
    price: '2.00',
  },
  {
    id: 3,
    name: 'Macchiato',
    description: '60ml espresso + a dab of milk foam',
    price: '2.20',
  },
];
const mockAddToOrder = jest.fn();
const mockRemoveFromOrder = jest.fn();

describe('Menu', () => {
  test('matches the snapshot', () => {
    const tree = renderer
      .create(
        <Menu
          stage={stage}
          order={order}
          menu={menu}
          addToOrder={mockAddToOrder}
          removeFromOrder={mockRemoveFromOrder}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
