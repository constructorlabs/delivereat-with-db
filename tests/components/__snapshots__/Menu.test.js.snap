// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu matches the snapshot 1`] = `
<div
  className="menu"
>
  <div
    className="menu-item menu-item--in-basket menu-item--greyed-out"
  >
    <div
      className="menu-item__title"
    >
      <div
        className="menu-item__name"
      >
        Espresso
      </div>
      <div
        className="menu-item__buttons"
      >
        <button
          className="menu-item__remove-button menu-item__remove-button--in-basket"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fas fa-minus-circle"
          />
        </button>
        <div
          className="menu-item__count menu-item__count--in-basket"
        >
          1
        </div>
        <button
          className="menu-item__add-button"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fas fa-plus-circle"
          />
        </button>
      </div>
    </div>
    <div
      className="menu-item__description"
    >
      Single shot, 30ml
    </div>
    <div
      className="menu-item__price"
    >
      £2.00
    </div>
  </div>
  <div
    className="menu-item menu-item--in-basket menu-item--greyed-out"
  >
    <div
      className="menu-item__title"
    >
      <div
        className="menu-item__name"
      >
        Ristretto
      </div>
      <div
        className="menu-item__buttons"
      >
        <button
          className="menu-item__remove-button menu-item__remove-button--in-basket"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fas fa-minus-circle"
          />
        </button>
        <div
          className="menu-item__count menu-item__count--in-basket"
        >
          3
        </div>
        <button
          className="menu-item__add-button"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fas fa-plus-circle"
          />
        </button>
      </div>
    </div>
    <div
      className="menu-item__description"
    >
      Concentrated espresso, 22ml
    </div>
    <div
      className="menu-item__price"
    >
      £2.00
    </div>
  </div>
  <div
    className="menu-item menu-item--greyed-out"
  >
    <div
      className="menu-item__title"
    >
      <div
        className="menu-item__name"
      >
        Macchiato
      </div>
      <div
        className="menu-item__buttons"
      >
        <button
          className="menu-item__remove-button"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fas fa-minus-circle"
          />
        </button>
        <div
          className="menu-item__count"
        >
          0
        </div>
        <button
          className="menu-item__add-button"
          onClick={[Function]}
          type="button"
        >
          <i
            className="fas fa-plus-circle"
          />
        </button>
      </div>
    </div>
    <div
      className="menu-item__description"
    >
      60ml espresso + a dab of milk foam
    </div>
    <div
      className="menu-item__price"
    >
      £2.20
    </div>
  </div>
</div>
`;
